### PR TITLE
Add scene and script entities

### LINKED ISSUE #63: Feature Request: Activation of scene and script entities
It would be great if it would be possible to activate scenes via your extension instead of controlling lights and switches directly

### PR DESCRIPTION
**Pull request by MexHigh**
Closes #63 

### State of this PR
- [x] The PR is currently based on master branch. I will try to rebase it to PR #62 once my stuff is working
- [x] Make sure, this works with script entities, too (currenty only working with scenes)
- [x] Add Translations for at least german
- [x] Add icons for scripts and scenes

### CODE REVIEW

**Comment by MexHigh** [comment]
Should work now again! Just need to do some code cleanup and add some translations.

Just to clarify the strange commit history: I wasn't able to rebase this many changes, so I made a backup of my changes inside my forked repo, reset the master branch to the upstream master and implemented my changes by hand again.

**Comment by geoph9** [comment]
Thank you! I'll look at it more thoroughly during the upcoming week (it'll take a bit more testing since I am not used to handling scenes).

**Comment by geoph9** [comment]
Since you only hit the `<URL>/api/services/scence/turn_on` endpoint for scenes/scripts, does that mean that we can only turn on those entities and not off? Or does hitting the `turn_on` endpoint twice result in the scene/script turning off?

**Comment by MexHigh** [comment]
Scene Entities don't have a `turn_off` service. The `turn_on` service simply applies the scene settings. They don't behave like a switch would.

A scripts execution can be stopped with a `turn_off` call, however, scripts are meant to be called once and then run until all actions are done. A call to `turn_off` is more like a forced cancellation for a script entity, which is why I don't think it is necessary to be implemented in a tool where you can "just click entities to run them".

**Comment by MexHigh** [comment]
Friedly ping :)

**Comment by geoph9** [comment]
Thanks for the reminder @MexHigh ! I had actually already reviewed the code and it works fine.

The only issue is that there is a lot of reusable code (i.e. functions whose body is 90% the same code). One way to solve this would be to add an extra level of abstraction (another superclass), or keep on using the functional approach and add further arguments to those functions (this may turn ugly, though). I'll try and take a look later today to see if this is easily solved.

**Comment by MexHigh** [comment]
Ok hit me up if there is something I can do or help you with

**Comment by geoph9** [comment]
Sorry I've been inactive lately. Could you maybe give it a try and remove the duplicated code? I.e. in many cases we have duplicated functions that only change e.g. a part of the url, or an error message. Ideally, we'd like to merge this kind of functionality into single functions.

**Comment by MexHigh** [comment]
Ok I'll look into it this week 👍🏼

File: utils.js [hunk 1]
```diff
@@ -612,25 +612,27 @@ const getMethods = (obj) => {
...
+        case "script":
+            icon_path += '/icons/script-text.svg';
+            break;
+        case "light":
+            icon_path += '/icons/ceiling-light.svg';
+            break;
+        case "fan":
+            icon_path += '/icons/fan.svg';
+            break;
+        case "switch":
```
Comment (geoph9) [inline utils.js:664]: What is the default path here?

File: utils.js [hunk 2]
```diff
@@ -612,25 +612,27 @@ const getMethods = (obj) => {
...
+        case "input_boolean":
+            icon_path += '/icons/toggle-switch-outline.svg';
+            break;
+        // no need for a default as these are all the supported domains by the plugin
```
Comment (geoph9) [inline utils.js:668]: Could you add a log message as the default? It could help debug potential future issues.

File: utils.js [hunk 3]
```diff
@@ -276,7 +276,7 @@ function invalidateEntitiesCache() {
  * @param {Boolean} force_reload Force reloading cache (optional, default: false)
  *
  */
-function getTogglables(callback, on_error=null, only_enabled=false, force_reload=false) {
```
Comment (geoph9) [inline utils.js:278]: You could completely delete this function.

File: utils.js [hunk 4]
```diff
@@ -305,7 +305,7 @@ function getTogglables(callback, on_error=null, only_enabled=false, force_reload
  * @param {Boolean} force_reload Force reloading cache (optional, default: false)
  *
  */
-function getRunnables(callback, on_error=null, only_enabled=false, force_reload=false) {
```
Comment (geoph9) [inline utils.js]: Same as with getTogglables

**Review by geoph9 — COMMENTED** [review]
Thanks! Seems good. Only those minor things need to be fixed and I'll merge it!

File: utils.js (hunk 1 above)
Comment (MexHigh) [reply utils.js:614]: None. All cases match the domains that can be produced by the `getEntities...` functions, so there will never an entity that does not match any of these domains.

File: utils.js (hunk 2 above)
Comment (MexHigh) [reply utils.js:618]: Yes, that's a good idea!

**Comment by MexHigh** [comment]
I guess I'm done. Please take another look at the changes in `prefs.js` after your review.

**Comment by geoph9** [comment]
Thanks @MexHigh, great work! I'll merge this now. I won't publish these changes on e.g.o. but I will create a release here.

On e.g.o. these changes will only work for Gnome > 45.

DISCLAIMER: If someone reaaally wants to use scripts and scenes with Gnome < 45 and (for some reason) does not want to use the releases page, then ping me here.

**Pull request merged**